crates/hir/src/semantics.rs

```diff
@@ -323,6 +323,61 @@ impl<'db> SemanticsImpl<'db> {
         tree
     }
 
+    pub fn find_parent_file(&self, file_id: HirFileId) -> Option<InFile<SyntaxNode>> {
+        match file_id.repr() {
+            HirFileIdRepr::FileId(file_id) => {
+                let module = self.file_to_module_defs(file_id.file_id()).next()?;
+                let def_map = self.db.crate_def_map(module.krate().id);
+                match def_map[module.id.local_id].origin {
+                    ModuleOrigin::CrateRoot { .. } => None,
+                    ModuleOrigin::File { declaration, declaration_tree_id, .. } => {
+                        let file_id = declaration_tree_id.file_id();
+                        let in_file = InFile::new(file_id, declaration);
+                        let node = in_file.to_node(self.db.upcast());
+                        let root = find_root(node.syntax());
+                        self.cache(root, file_id);
+                        Some(in_file.with_value(node.syntax().clone()))
+                    }
+                    _ => unreachable!("FileId can only belong to a file module"),
+                }
+            }
+            HirFileIdRepr::MacroFile(macro_file) => {
+                let node = self
+                    .db
+                    .lookup_intern_macro_call(macro_file.macro_call_id)
+                    .to_node(self.db.upcast());
+                let root = find_root(&node.value);
+                self.cache(root, node.file_id);
+                Some(node)
+            }
+        }
+    }
+
+    /// Returns the `SyntaxNode` of the module. If this is a file module, returns
+    /// the `SyntaxNode` of the *definition* file, not of the *declaration*.
+    pub fn module_definition_node(&self, module: Module) -> InFile<SyntaxNode> {
+        let def_map = module.id.def_map(self.db.upcast());
+        match def_map[module.id.local_id].origin {
+            ModuleOrigin::CrateRoot { definition } => {
+                InFile::new(definition.into(), self.parse(definition).syntax().clone())
+            }
+            ModuleOrigin::File { definition, .. } => {
+                InFile::new(definition.into(), self.parse(definition).syntax().clone())
+            }
+            ModuleOrigin::Inline { definition_tree_id, definition } => {
+                let file_id = definition_tree_id.file_id();
+                let in_file = InFile::new(file_id, definition);
+                let node = in_file.to_node(self.db.upcast());
+                let root = find_root(node.syntax());
+                self.cache(root, file_id);
+                in_file.with_value(node.syntax().clone())
+            }
+            ModuleOrigin::BlockExpr { .. } => {
+                panic!("Semantics::module_definition_node() doesn't handle block modules")
```

Comment:
Can we just return `ModuleSource` like `ModuleOrigin::definition_source` does and filter at the call site?